### PR TITLE
OLS-447: LLM parameters mapping

### PR DESCRIPTION
**Pull request by tisnik**
## Description

[OLS-447](https://issues.redhat.com//browse/OLS-447): LLM parameters mapping

## Type of change

- [ ] Refactor
- [x] New feature
- [ ] Bug fix
- [ ] CVE fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Configuration Update
- [ ] Bump-up dependent library

## Related Tickets & Documents

- Related Issue #[OLS-447](https://issues.redhat.com//browse/OLS-447)
- Closes #

### CODE REVIEW

**Comment by tisnik** [comment]
/assign @bparees 

This is not finished (missing tests etc.), but I'd like to have feedback if we go in the right direction.

**Comment by bparees** [comment]
> This is not finished (missing tests etc.), but I'd like to have feedback if we go in the right direction.

yes the direction seems right to me.

File: ols/src/llms/providers/provider.py [hunk 1]
```diff
@@ -89,6 +92,32 @@ class ProviderParameter:
...
+# Generic to BAM parameters mapping
+BAMParametersMapping: dict[str, str] = {}
+
+# Generic to Watsonx parameters mapping
+WatsonxParametersMapping: dict[str, str] = {
+    "min_new_tokens": GenParams.MIN_NEW_TOKENS,
+    "max_new_tokens": GenParams.MAX_NEW_TOKENS,
+    "top_k": GenParams.TOP_K,
+    "top_p": GenParams.TOP_P,
+    "temperature": GenParams.TEMPERATURE,
```
Comment (bparees) [inline ols/src/llms/providers/provider.py:110]: if these are to be our official "generic llm parameter names", should we make them into constants?

File: ols/src/query_helpers/question_validator.py [hunk 2]
```diff
@@ -21,9 +21,8 @@ class QuestionValidator(QueryHelper):
 
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         """Initialize the QuestionValidator."""
-        # TODO: OLS-447 Refactor code that pass parameters into LLM in QuestionValidator
-        llm_params = {"max_new_tokens": 4}
-        super().__init__(*args, **dict(kwargs, llm_params=llm_params))
+        generic_llm_params = {"max_new_tokens": 4}
```
Comment (bparees) [inline ols/src/query_helpers/question_validator.py:24]: replace w/ generic parameter constant if you agree to creating one.

File: tests/unit/llms/providers/test_watsonx.py [hunk 3]
```diff
@@ -142,3 +142,38 @@ def test_params_replace_default_values_with_none(provider_config):
...
+    """Test generic parameter mapping to provider parameter list."""
+    config.init_empty_config()  # needed for checking the config.dev_config.llm_params
+
+    # some non default values
+    generic_llm_params = {
+        "min_new_tokens": 100,
+        "max_new_tokens": 200,
+        "top_k": 10,
+        "top_p": 1.5,
+        "temperature": 42.0,
```
Comment (bparees) [inline tests/unit/llms/providers/test_watsonx.py:158]: use constants if you agree to create them

File: tests/unit/query_helpers/test_question_validator.py [hunk 4]
```diff
@@ -20,16 +23,35 @@ def test_is_query_helper_subclass():
...
+    """Test that generic_llm_params is handled correctly and without runtime error."""
     # it is needed to initialize configuration in order to be able
     # to construct QuestionValidator instance
     config.init_config("tests/config/valid_config.yaml")
 
     question_validator = QuestionValidator()
-    assert question_validator.llm_params is not None
-    assert "max_new_tokens" in question_validator.llm_params
+    assert question_validator.generic_llm_params is not None
+    assert "max_new_tokens" in question_validator.generic_llm_params
```
Comment (bparees) [inline tests/unit/query_helpers/test_question_validator.py:33]: constant

File: tests/unit/query_helpers/test_question_validator.py [hunk 5]
```diff
@@ -20,16 +23,35 @@ def test_is_query_helper_subclass():
...
+
+    question_validator = QuestionValidator(generic_llm_params={})
+    # the generic_llm_params should be rewritten in constructor
+    assert question_validator.generic_llm_params is not None
+    assert "max_new_tokens" in question_validator.generic_llm_params
```
Comment (bparees) [inline tests/unit/query_helpers/test_question_validator.py:38]: constant

File: tests/unit/query_helpers/test_question_validator.py [hunk 6]
```diff
@@ -20,16 +23,35 @@ def test_is_query_helper_subclass():
...
+
+@patch("ols.src.query_helpers.question_validator.LLMChain", new=mock_llm_chain(None))
+def test_validate_question_llm_loader():
+    """Test that LLM is loaded within validate_question method with proper parameters."""
+    # it is needed to initialize configuration in order to be able
+    # to construct QuestionValidator instance
+    config.init_config("tests/config/valid_config.yaml")
+
+    llm_loader = mock_llm_loader(
+        None, expected_params=("p1", "m1", {"max_new_tokens": 4})
```
Comment (bparees) [inline tests/unit/query_helpers/test_question_validator.py:49]: constant

File: tests/unit/query_helpers/test_question_validator.py [hunk 7]
```diff
@@ -20,16 +23,35 @@ def test_is_query_helper_subclass():
...
+    )
 
-    question_validator = QuestionValidator(llm_params={})
-    # the llm_params should be rewritten in constructor
-    assert question_validator.llm_params is not None
-    assert "max_new_tokens" in question_validator.llm_params
+    # check that LLM loader was called with expected parameters
```
Comment (bparees) [inline tests/unit/query_helpers/test_question_validator.py:52]: how is this actually verifying that it was called w/ the expected parameters?

File: tests/unit/query_helpers/test_question_validator.py [hunk 8]
```diff
@@ -20,16 +23,35 @@ def test_is_query_helper_subclass():
...
+    question_validator = QuestionValidator(llm_loader=llm_loader)
+    validated = question_validator.validate_question(
+        "123e4567-e89b-12d3-a456-426614174000", "query"
+    )
+    assert validated
```
Comment (bparees) [inline tests/unit/query_helpers/test_question_validator.py:57]: be explicit about what you're asserting (`validated is True`).  If we later changed question_validator to return a string, this assert would pass for any non-empty value.

but also i'm not sure why we are checking the return value of question_validator here, it doesn't seem to be the focus of the test.

**Review by bparees — COMMENTED** [review]
variety of nits

File: ols/src/llms/providers/provider.py (hunk 1 above)
Comment (tisnik) [reply ols/src/llms/providers/provider.py:110]: yup

File: ols/src/query_helpers/question_validator.py (hunk 2 above)
Comment (tisnik) [reply ols/src/query_helpers/question_validator.py:24]: +1

Comment (tisnik) [reply ols/src/query_helpers/question_validator.py]: updated @bparees

File: tests/unit/llms/providers/test_watsonx.py (hunk 3 above)
Comment (tisnik) [reply tests/unit/llms/providers/test_watsonx.py]: updated @bparees

File: tests/unit/query_helpers/test_question_validator.py (hunk 4 above)
Comment (tisnik) [reply tests/unit/query_helpers/test_question_validator.py]: also updated @bparees

File: tests/unit/query_helpers/test_question_validator.py (hunk 5 above)
Comment (tisnik) [reply tests/unit/query_helpers/test_question_validator.py]: dtto

File: tests/unit/query_helpers/test_question_validator.py (hunk 7 above)
Comment (tisnik) [reply tests/unit/query_helpers/test_question_validator.py:61]: this is the trick:
```
llm_loader = mock_llm_loader(
        None, expected_params=("p1", "m1", {GenericLLMParameters.MAX_NEW_TOKENS: 4})
    )
```

I've added comment to make it a bit more clear

File: tests/unit/query_helpers/test_question_validator.py (hunk 8 above)
Comment (tisnik) [reply tests/unit/query_helpers/test_question_validator.py]: updated a bit

**Pull request merged**